### ISSUE TITLE
Bug: OutOfMemoryError on Android

### ISSUE THREAD

**Issue by tttppp** [andybalaam/rabbit-escape#116]
When playing K2 (v0.6.1) on Android I consistently get a crash with the following stacktrace:

```
E/art     (17831): Throwing OutOfMemoryError "Failed to allocate a 157476 byte allocation with 138064 free bytes and 134KB until OOM"
E/AndroidRuntime(17831): FATAL EXCEPTION: GameLaunch
E/AndroidRuntime(17831): Process: net.artificialworlds.rabbitescape, PID: 17831
E/AndroidRuntime(17831): java.lang.OutOfMemoryError: Failed to allocate a 157476 byte allocation with 138064 free bytes and 134KB until OOM
E/AndroidRuntime(17831):        at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
E/AndroidRuntime(17831):        at android.graphics.Bitmap.nativeCreate(Native Method)
E/AndroidRuntime(17831):        at android.graphics.Bitmap.createBitmap(Bitmap.java:939)
E/AndroidRuntime(17831):        at android.graphics.Bitmap.createBitmap(Bitmap.java:912)
E/AndroidRuntime(17831):        at android.graphics.Bitmap.createBitmap(Bitmap.java:843)
E/AndroidRuntime(17831):        at android.graphics.Bitmap.createScaledBitmap(Bitmap.java:719)
E/AndroidRuntime(17831):        at rabbitescape.ui.android.AndroidBitmapScaler.scale(AndroidBitmapScaler.java:21)
E/AndroidRuntime(17831):        at rabbitescape.ui.android.AndroidBitmapScaler.scale(AndroidBitmapScaler.java:9)
E/AndroidRuntime(17831):        at rabbitescape.render.ScaledBitmap.scaleTo(ScaledBitmap.java:69)
E/AndroidRuntime(17831):        at rabbitescape.render.ScaledBitmap.bitmap(ScaledBitmap.java:36)
E/AndroidRuntime(17831):        at rabbitescape.render.Renderer.drawSprite(Renderer.java:35)
E/AndroidRuntime(17831):        at rabbitescape.render.Renderer.render(Renderer.java:28)
E/AndroidRuntime(17831):        at rabbitescape.ui.android.AndroidGraphics.drawToCanvas(AndroidGraphics.java:227)
E/AndroidRuntime(17831):        at rabbitescape.ui.android.AndroidGraphics.actuallyDrawGraphics(AndroidGraphics.java:168)
E/AndroidRuntime(17831):        at rabbitescape.ui.android.AndroidGraphics.draw(AndroidGraphics.java:124)
E/AndroidRuntime(17831):        at rabbitescape.render.gameloop.GameLoop.step(GameLoop.java:48)
E/AndroidRuntime(17831):        at rabbitescape.render.gameloop.GameLoop.run(GameLoop.java:30)
E/AndroidRuntime(17831):        at rabbitescape.ui.android.AndroidGameLaunch.run(AndroidGameLaunch.java:73)
E/AndroidRuntime(17831):        at java.lang.Thread.run(Thread.java:818)
W/ActivityManager( 9958):   Force finishing activity net.artificialworlds.rabbitescape/rabbitescape.ui.android.AndroidGameActivity
```

**Comment by andybalaam** [comment]
Sounds like the same thing I saw on my S3. I can reproduce by zooming in on most levels. My plan is to catch out of memory exceptions and zoom out and/or reduce the size of the image cache. Better ideas welcome.

**Comment by colonelfazackerley** [comment]
Were you going to speed it up by only over painting in regions of movement? If the image cache only covered changes then this should save memory too.

**Comment by andybalaam** [comment]
I guess so but really the cache fills up when you see varied animation rather than drawing lots of things simultaneously. We could just make the cache size small by default but that would slow us down on larger devices or just when we're less zoomed in.

**Comment by andybalaam** [comment]
A couple more stack traces from real users :-(

    java.lang.OutOfMemoryError
	at android.graphics.Bitmap.nativeCreate(Native Method)
	at android.graphics.Bitmap.createBitmap(Bitmap.java:718)
	at android.graphics.Bitmap.createBitmap(Bitmap.java:695)
	at android.graphics.Bitmap.createBitmap(Bitmap.java:628)
	at android.graphics.Bitmap.createScaledBitmap(Bitmap.java:508)
	at rabbitescape.ui.android.AndroidBitmapScaler.scale(AndroidBitmapScaler.java:21)
	at rabbitescape.ui.android.AndroidBitmapScaler.scale(AndroidBitmapScaler.java:9)
	at rabbitescape.render.ScaledBitmap.scaleTo(ScaledBitmap.java:69)
	at rabbitescape.render.ScaledBitmap.bitmap(ScaledBitmap.java:36)
	at rabbitescape.render.Renderer.drawSprite(Renderer.java:35)
	at rabbitescape.render.Renderer.render(Renderer.java:28)
	at rabbitescape.ui.android.AndroidGraphics.drawToCanvas(AndroidGraphics.java:227)
	at rabbitescape.ui.android.AndroidGraphics.actuallyDrawGraphics(AndroidGraphics.java:168)
	at rabbitescape.ui.android.AndroidGraphics.draw(AndroidGraphics.java:124)
	at rabbitescape.render.gameloop.GameLoop.step(GameLoop.java:48)
	at rabbitescape.render.gameloop.GameLoop.run(GameLoop.java:30)
	at rabbitescape.ui.android.AndroidGameLaunch.run(AndroidGameLaunch.java:73)
	at java.lang.Thread.run(Thread.java:856)

    java.lang.OutOfMemoryError
	at android.graphics.BitmapFactory.nativeDecodeAsset(Native Method)
	at android.graphics.BitmapFactory.decodeStream(BitmapFactory.java:677)
	at android.graphics.BitmapFactory.decodeStream(BitmapFactory.java:717)
	at rabbitescape.ui.android.AndroidBitmapLoader.load(AndroidBitmapLoader.java:38)
	at rabbitescape.ui.android.AndroidBitmapLoader.load(AndroidBitmapLoader.java:15)
	at rabbitescape.render.ScaledBitmap.scaleTo(ScaledBitmap.java:57)
	at rabbitescape.render.ScaledBitmap.bitmap(ScaledBitmap.java:36)
	at rabbitescape.render.Renderer.drawSprite(Renderer.java:35)
	at rabbitescape.render.Renderer.render(Renderer.java:28)
	at rabbitescape.ui.android.AndroidGraphics.drawToCanvas(AndroidGraphics.java:227)
	at rabbitescape.ui.android.AndroidGraphics.actuallyDrawGraphics(AndroidGraphics.java:168)
	at rabbitescape.ui.android.AndroidGraphics.draw(AndroidGraphics.java:124)
	at rabbitescape.render.gameloop.GameLoop.step(GameLoop.java:48)
	at rabbitescape.render.gameloop.GameLoop.run(GameLoop.java:30)
	at rabbitescape.ui.android.AndroidGameLaunch.run(AndroidGameLaunch.java:73)
	at java.lang.Thread.run(Thread.java:841)

**Comment by andybalaam** [comment]
Feeling really bad about this.  As soon as the arcade levels release is out, this is what I will focus on.

**Comment by andybalaam** [comment]
The plan:

1. Catch out-of-memory errors somewhere like Renderer.drawSprite, and if we get one, shrink the cache (without using any memory!) and try again.  Also, possibly, zoom out, if you are zoomed in.

2. Modify the cache to hold only 1 copy of each image, instead of a scaled and unscaled copy.  (Optimise for memory instead of performance)

3. Shrink the default cache size (and possibly limit the max zoom level) so that this doesn't trigger for anyone we know.

Any other suggestions?

**Comment by andybalaam** [comment]
I wonder whether we can run graphical system tests with small heap sizes passed in to the JVM to simulate low-memory conditions...

**Comment by andybalaam** [comment]
(and possibly something similar on Android)

**Comment by andybalaam** [comment]
Thinking a bit more, I also suspect even when we're zoomed in to a small area of level we still load and scale bitmaps for everything that is happening.

**Comment by andybalaam** [comment]
I'm going to try and work on this tonight.

**Comment by andybalaam** [comment]
I've been working on a cache that is similar to Android's [LruCache](https://developer.android.com/reference/android/util/LruCache.html) (useful link: [Caching Bitmaps](https://developer.android.com/training/displaying-bitmaps/cache-bitmap.html) ).

I decided not to use the Android LruCache because a) it's not available on API level 8 and you have to install a support package to use it and b) we need the same thing in Swing.

**Comment by andybalaam** [comment]
I've realised that the ScaledBitmap + BitmapCache setup I wrote is not working out brilliantly, because it loads bitmaps lazily, and that means you don't know their size when you put them in the cache.  In addition, the system I wrote that holds an unscaled image in memory and provides scaled versions on demands, is an unproven ahead-of-time optimisation that I now suspect is a very bad trade-off, prioritising speed over memory usage.

**Comment by andybalaam** [comment]
I am working on changing this, but it won't get done tonight.

**Comment by andybalaam** [comment]
Sorry, we support API level 9 (version 2.2), not 8.

**Comment by andybalaam** [comment]
The work I've done so far, which measures the size of the currently cached bitmaps, and removes some if we run out of space, may already fix this bug.  I am no longer able to reproduce it on my phone.

**Comment by andybalaam** [comment]
TODO:

1. Get rid of ScaledBitmap.  Instead, make BitmapCache scale on demand, and store each bitmap in the LruCache under a name like "filename-32", so when we need different sizes we create and store different images under different cache keys.

2. Consider clearing the cache when we zoom.

**Comment by andybalaam** [comment]
I've uploaded a snapshot (version 0.7.0.1) to http://artificialworlds.net/rabbit-escape/snapshots/ - @tttppp please could you try and crash it?  What I've done so far is very crude, and I think loading lots of images when small, then zooming in, may crash it, but I'd be interested to hear whether it's any better at all.

**Comment by tttppp** [comment]
Looks good now. I can no longer crash it by simply playing K2 or Tetris at a zoomed in view.

**Comment by andybalaam** [comment]
Brilliant I did actually manage to crash it after I posted that, but it does seem much harder. Once I've finished it should be effectively impossible.

**Comment by andybalaam** [comment]
The only tunable parameter left when I'm done will be what proportion of memory is given over to the image cache. I am going with half but we'll have to see how that goes.

**Comment by andybalaam** [comment]
I've done more work on this, and I think I have a proper fix - the size of bitmaps is fixed when they are put in the cache, so it should not be possible to trick it into using more memory than it should.  I've got rid of the whole ScaledBitmap structure, and just request the right-size bitmap from the cache in the Renderer.

But it's was quite a big change, so I'd like to have lots of testing before we release anything.

I've put a new snapshot (0.7.0.2) here: http://artificialworlds.net/rabbit-escape/snapshots/

**Comment by andybalaam** [comment]
TODOs from above:

1. is done
2. I think we won't do - if the user zooms in then out, we may as well leave the previous zoom levels in the cache if there's free memory, so zooming back out it faster
3. is done
4. is done because ScaledBitmap is gone
5. is ongoing - please help test.

new TODO:

6. Consider whether we can avoid loading bitmaps when they are off screen.  This would be easy for bitmaps that are off the right or bottom.  To do ones off the top or left we would need to know how big they are without loading them.  We could do some kind of preprocessing during the build to find this out.  I'm not inclined to do this yet...

**Comment by tttppp** [comment]
Did a fair bit of testing today with latest version 0.7.0.2. No crashes, so I vote this should replace the current Play Store version.

**Comment by andybalaam** [comment]
I managed to crash 0.7.0.2 a few times. My theory was that we were leaking cache objects themselves and I've made the cache a Singleton spit spit and so far memory usage looks really constrained and l can't crash it. I'll get it working properly and make another snapshot. Thanks for all the testing, appreciated.

**Comment by andybalaam** [comment]
Wanted to get a new version out for Christmas but hey I want to make it right.

**Comment by andybalaam** [comment]
New snapshot 0.7.0.3 uploaded to http://artificialworlds.net/rabbit-escape/snapshots/ . I think this one is the one.

**Comment by andybalaam** [comment]
Fixed by commit 610ab30 and the preceding few.  Opened a new issue #229 for the separate problem of not loading images we don't need.

If you see any out of memory errors in 0.7.1 or beyond I am very interested in them, since I think this should now be fixed.
